doc: ci: workflow improvements and PDF support

gmarull opened the pull request:
This PR brings some improvements to the documentation workflow:

- A single "Documentation Build" workflow is used to build the documentation, regardless if it's for a PR, a release or a scheduled event.
- The "Documentation Publish" workflow will run on `main` or releases after documentation build succeeds. It re-uses artifacts from the build stage.
- PDF support has been added (build and publish)

NOTE: I'm not entirely familiar with how AWS S3 is configured for Zephyr, so that part may require more attention as I may have missed something.

gmarull committed:
ci: doc: add support for PDF build

Add support for building the PDF version of the documentation.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

nashif commented on .github/workflows/doc-build.yml:
```diff
@@ -63,3 +63,43 @@ jobs:
+
+    - name: build-docs
+      run: |
+        cmake -Sdoc -Bbuild -DSPHINXOPTS="-q -j auto" -DLATEXMKOPTS="-quiet -halt-on-error"
```
why not call 'make pdf' to make sure our build process also works?

gmarull replied:
added `LATEXMKOPTS` option to Makefile (required for CI), now CI uses "make pdf"

nashif commented on .github/workflows/doc-build.yml:
```diff
@@ -63,3 +63,43 @@ jobs:
+      uses: actions/upload-artifact@master
+      continue-on-error: True
+      with:
+        path: build/latex/zephyr.pdf
```
add name here, so it is visible that this is the PDF.

```
name: pdf-output
```

for example.

gmarull replied:
added

gmarull committed:
doc: add link to download PDF file

Add a new downloads section with a link to the Zephyr PDF docs.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: makefile: add support for latexmk options

Allow to override latexmk options via Makefile. Defaults match those in
the CMakeLists.txt.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: use xz compression for html archive

xz compression reduces the html archive sice from ~350MB to ~90MB (at
the cost of more compression time). Compression has been moved to a
separate step.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: do not let west init fail

Modules are required to generate, for example, full Kconfig docs.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: use the hash of requirements file as cache key

The pip cache should be discarded if any of the required dependencies
changes, so use the hash of the requirements file as a key source.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: remove unnecessary west path update step

The step is not necessary to build the docs correctly, so remove it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: run doc build on schedule and releases

The documentation build workflow is re-used for the periodic build and
releases.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: use an env var to store west version

Storing west version in a single place avoids keeping duplicates in each
job. Also added a note on why pinning version is important: west
docstrings are extracted from the installed version.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: do not succeed if artifact upload fails

Artifacts will be required for the publish workflow, so do not let the
build workflow succeed without them being uploaded correctly.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: set doc tag when building

Set the correct dog tag based on the Github event name.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: refactor doc publish

Publish action has been refactored into a workflow_run based action.
The action is now triggered after a successful doc-build workflow run
(only on main/release branches). The artifacts are downloaded from the
build run and published accordingly. Publish code has been simplified a
bit and PDF upload has been added.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc: adjust build workflow name

Adjust the name to "Documentation Build".

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

carlescufi reviewed (approved):
Looks good to me!

nashif commented on .github/workflows/doc-build.yml:
```diff
@@ -4,6 +4,11 @@
 name: Documentation GitHub Workflow
 
 on:
+  schedule:
```
do not understand why this is needed.You need to explain why do we need another job building the docs on schedule in the main branch, we have this already.

gmarull replied:
The publish workflow has been refactored to use the workflow_run event. It doesn't build anymore, just publishes from imported artifacts. This way we have a single workflow that is responsible for building docs and so we remove duplicates. A working example: https://github.com/nrfconnect/sdk-nrf/blob/master/.github/workflows/docpublish.yml

nashif commented on .github/workflows/doc-publish.yml:
```diff
@@ -1,48 +1,35 @@
 # SPDX-License-Identifier: Apache-2.0
 
-name: Doc build for Release or Daily
+name: Documentation Publish
```
Publish Documentation

gmarull replied:
adjusted

mbolivar-nordic commented on .github/workflows/doc-build.yml:
```diff
@@ -25,10 +25,6 @@ jobs:
     steps:
-    - name: Update PATH for west
-      run: |
-        echo "$HOME/.local/bin" >> $GITHUB_PATH
```
I guess recent ubuntu is doing this for the user automatically? It was required on that OS at some point in the past.

gmarull replied:
This is the `PATH` of the `ubuntu-latest` runner:

```
PATH=/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```

so yes, `/home/runner/.local/bin` is already there.

mbolivar-nordic reviewed (approved):
Sure, no objections to this approach
